### ISSUE TITLE
Crash if sysfs network speed pseudofile is unavailable

### ISSUE THREAD

**Issue by shivramsrivastava** [camsas/firmament#48]
We have Ubuntu VM's Created on OpenStack.
The build went fine, but the coordinator crashes with this error message.
Because this file '/sys/class/net/eth0/speed' has no data and is read by the function 'GetMachineCapacity'.

The following is the error message:

```
ubuntu:~/workspace/src/firmament$ sudo build/src/coordinator --listen_uri tcp:10.11.12.111:9091 --task_lib_dir=$(pwd)/build/src/
F1006 17:54:10.509845  1742 procfs_machine.h:64] Check failed: fscanf(input, "%ju ", x) == 1 (-1 vs. 1)
*** Check failure stack trace: ***
    @     0x7f57273fedaa  (unknown)
    @     0x7f57273fece4  (unknown)
    @     0x7f57273fe6e6  (unknown)
    @     0x7f5727401687  (unknown)
    @           0x60c2e4  firmament::platform_unix::ProcFSMachine::readunsigned()
    @           0x60d4c4  firmament::platform_unix::ProcFSMachine::GetMachineCapacity()
    @           0x56790c  firmament::Coordinator::AddResource()
    @           0x5fea8e  firmament::BFSTraverseResourceProtobufTreeReturnRTND()
    @           0x5680d9  firmament::Coordinator::DetectLocalResources()
    @           0x56d5f1  firmament::Coordinator::Run()
    @           0x55e80f  main
    @     0x7f5725ebcf45  (unknown)
    @           0x5641c0  (unknown)
    @              (nil)  (unknown)
Aborted (core dumped)
```

Relevant source code:

``` c++
void ProcFSMachine::GetMachineCapacity(ResourceVector* cap) {
  // Extract the total available resource capacities on this machine
  MemoryStatistics_t mem_stats = GetMemoryStats();
  cap->set_ram_cap(mem_stats.mem_total / BYTES_TO_MB);
  vector<CPUStatistics_t> cpu_stats = GetCPUStats();
  // Subtract one as we have an additional element for the overall CPU load
  // across all cores
  cap->set_cpu_cores(cpu_stats.size() - 1);
  // Get network interface speed from ProcFS
  string nic_speed_path;
  spf(&nic_speed_path, "**/sys/class/net/%s/speed**",
      FLAGS_monitor_netif.c_str());
  FILE* nic_speed_fd = fopen(nic_speed_path.c_str(), "r");
  uint64_t speed = 0;
  if (nic_speed_fd) {
    **readunsigned(nic_speed_fd, &speed);             ---> fails here during the CHECK_EQ call..**
    CHECK_EQ(fclose(nic_speed_fd), 0);
  }
```

Is this an issue with our machines or a problem with the firmament?

I did the following workaround, and it seems to work fine.  But this 'readunsigned' function is called from a couple of other places, looks like not the correct fix, but a temporary workaround.

``` diff
   inline void readunsigned(FILE* input, uint64_t *x) {
-    CHECK_EQ(fscanf(input, "%ju ", x), 1);
+    -fscanf(input, "%ju ", x);
   }
```

**Comment by ms705** [comment]
Hi @shivramsrivastava,

Thanks for reporting this! The issue you're experiencing is likely related to the name of the network device on your OpenStack VM. By default, Firmament assumes that your primary network interface is called `eth0`, but this is often not the case on VMs.

You can change the interface name Firmament uses by setting the `--monitor_netif` flag to the right interface name. For example, if your interface was called `em0`, you would set `--monitor_netif=em0`.

Another possible explanation is that you don't have permission to read from procfs, but this shouldn't be the case in a VM (it might be in a non-privileged container). To get more information, you can patch the `CHECK_EQ` as follows:
```diff
   inline void readunsigned(FILE* input, uint64_t *x) {
-    CHECK_EQ(fscanf(input, "%ju ", x), 1);
-    CHECK_EQ(fscanf(input, "%ju ", x), 1) << "Failed to read unsigned int: " << strerror(errno);
   }
```
(This may require `#include <errno.h>`. `fscanf` sets `errno` rather than returning an indication of the specific error that occured.)

That said, we might want to either add auto-detection of network interface names (monitoring all interfaces by default), or make this monitoring optional.

**Comment by shivramsrivastava** [comment]
This is the error message we are getting.
Looks like we don't have read permission to some files in the 'sysfs' files system.
 
sudo build/src/coordinator --listen_uri tcp:10.11.12.111:9091 --task_lib_dir=$(pwd)/build/src/
F1013 07:37:29.912360 23158 procfs_machine.h:64] Check failed: fscanf(input, "%ju ", x) == 1 (-1 vs. 1) **Failed to read unsigned int: Invalid argument**
*** Check failure stack trace: ***
    @     0x7fdbba8d3daa  (unknown)
    @     0x7fdbba8d3ce4  (unknown)
    @     0x7fdbba8d36e6  (unknown)
    @     0x7fdbba8d6687  (unknown)
    @           0x60e592  firmament::platform_unix::ProcFSMachine::readunsigned()
    @           0x60d539  firmament::platform_unix::ProcFSMachine::GetMachineCapacity()
    @           0x5679bc  firmament::Coordinator::AddResource()
    @           0x5feb3e  firmament::BFSTraverseResourceProtobufTreeReturnRTND()
    @           0x568189  firmament::Coordinator::DetectLocalResources()
    @           0x56d6a1  firmament::Coordinator::Run()
    @           0x55e8bf  main
    @     0x7fdbb9391f45  (unknown)
    @           0x564270  (unknown)
    @              (nil)  (unknown)



These are the three net interface listed on our machine:

/sys/class/net$ ls
docker0  eth0  lo

Even with 'ethtool', information about the net interface is not available. 

~:/sys/class/net$ sudo ethtool eth0
Settings for eth0:
        Link detected: yes
~:/sys/class/net$ sudo ethtool -v eth0
ethtool: bad command line argument(s)
For more information run ethtool -h
~:/sys/class/net$ sudo ethtool -c eth0
Coalesce parameters for eth0:
Cannot get device coalesce settings: Operation not supported
~:/sys/class/net$ sudo ethtool -C eth0
Cannot get device coalesce settings: Operation not supported
~:/sys/class/net$ sudo ethtool -S eth0
no stats available

Also in the past, we have seen that some hardware information like the RAPL cannot be read unless we did some tweaks in the BIOS setting, does firmament require such similar tweaks to read hardware info?

**Comment by ms705** [comment]
Hi @shivramsrivastava,

No, BIOS changes should not be required; the `sysfs` permissions are purely a Linux software setting, and what happens when reading or writing the files in `sysfs` is driver-specific.

I've managed to reproduce your issue in an OpenStack VM, and it looks to me as though it isn't a permission issue (anyone can read `/sys/class/net/eth0/speed`), but rather a problem with the RedHat virtual NIC driver, which does not expose a NIC speed. (In a way, this isn't surprising: it's not totally clear what the "speed" of a virtual NIC is -- it really depends on the underlying physical NIC's speed, and on whether the virtual NIC talks to another local VM or to a remote destination!)

I think the right medium-term fix for this issue is this:
 1. Remove the `CHECK_EQ` assertion from `readunsigned`, allowing it to fail.
 2. Change the signature of `readunsigned` to return a signed int as an error code, set to 0 if an unsigned int was read into the second argument.
 3. Add assertions and error handling code around all calls to `readunsigned`.
 4. In the case of the network speed read code, add error handling code that assumes a sensible default (taken from a flag similar to `--monitor_blockdev_maxbw`) if no speed can be determined.

In the long term, it would be ideal if Firmament didn't have to itself poke around in `procfs` and `sysfs` to determine these things: instead, it would make sense to integrate it with a tool like [cAdvisor](https://github.com/google/cadvisor) or [snap](https://github.com/intelsdi-x/snap) by accessing information via their APIs. However, this is a much larger change.

Would you be able to make these changes and drop them into a pull request? I can also have a go myself, but it'll probably take a few days until I get around to this.

**Comment by ms705** [comment]
PR #50 has a solution to this problem, currently under review.

**Comment by ms705** [comment]
Fixed in 3aeb8d4094c6805f8b904b777c54ac349e491df4, so closing.
